File: src/main/kotlin/com/example/features/ClassFeaturesDemo.kt
package com.example.features

/**
 * Demonstrates class features not covered by Calculator / Animal / Shapes:
 *   - Class body properties (non-constructor backing fields)
 *   - Explicit property getter (get() = expression)
 *   - Explicit property setter (set(v) { this.field = v })
 *   - Interface stub emission
 *   - `when` with `is` type-pattern conditions
 */

/** Class with a non-constructor backing field mutated via instance methods. */
class Counter {
    var count: Int = 0

    fun increment(): Int {
        this.count++
        return this.count
    }

    fun reset() {
        this.count = 0
    }

    fun readCount(): Int = this.count
}

/** Class with explicit property getters (expression bodies). */
class Rectangle(val w: Int, val h: Int) {
    val area: Int
        get() = this.w * this.h

    val perimeter: Int
        get() = (this.w + this.h) * 2
}

/** Class with a var property and an explicit setter that writes through `this`. */
class Label(initial: String) {
    var text: String = initial
        set(v) { this.text = v }

    fun readText(): String = this.text
}

/**
 * Interface – emitted as a Lua table with stub functions.
 * Classes that implement it provide their own bodies.
 */
interface Describable {
    fun describe(): String
    fun shortName(): String
}

/**
 * Class that satisfies the Describable contract.
 * `describe()` uses a two-field string template to exercise
 * `tostring(self.field) .. "…" .. tostring(self.field)` codegen.
 */
class Product(val name: String, val price: Int) : Describable {
    override fun describe(): String = "${this.name}: ${this.price}"
    override fun shortName(): String = this.name
}

class ClassFeaturesDemo

/** `when` with `is` type-pattern conditions (String / Int / Boolean / else). */
fun demoTypeWhen(x: Any): String = when (x) {
    is String -> "string"
    is Int -> "number"
    is Boolean -> "boolean"
    else -> "other"
}

/** Returns the count after 3 increments. */
fun demoCounter(): Int {
    val c = Counter()
    c.increment()
    c.increment()
    c.increment()
    return c.readCount()   // 3
}

File: outputDir/vscripts/features/ClassFeaturesDemo.lua
-- package: com.example.features

require("ktox-lib")
ktox_sourcemap_traceback(debug and debug.getinfo and (debug.getinfo(1) or {}).short_src or "", "features/ClassFeaturesDemo.kt", {["1-15"]=1,["16"]=17,["17-20"]=18,["21-115"]=22,["116"]=76,["117"]=77,["118"]=78,["119"]=79,["120-122"]=80})

Counter = {}
Counter.__index = Counter

function Counter:new()
    local self = setmetatable({}, Counter)
    self.count = 0
    return self
end

function Counter:increment()
    self.count = self.count + 1
    return self.count
end

function Counter:reset()
    self.count = 0
end

function Counter:readCount()
    return self.count
end


Rectangle = {}
Rectangle.__index = Rectangle

function Rectangle:new(w, h)
    local self = setmetatable({}, Rectangle)
    self.w = w
    self.h = h
    return self
end

function Rectangle:getArea()
    return self.w * self.h
end

function Rectangle:getPerimeter()
    return (self.w + self.h) * 2
end


Label = {}
Label.__index = Label

function Label:new(initial)
    local self = setmetatable({}, Label)
    self.text = initial
    return self
end

function Label:setText(v)
    self.text = v
end

function Label:readText()
    return self.text
end


-- Interface: Describable
local Describable = {}
Describable.__index = Describable

function Describable:describe() end
function Describable:shortName() end

Product = {}
Product.__index = Product

function Product:new(name, price)
    local self = setmetatable({}, Product)
    self.name = name
    self.price = price
    return self
end

function Product:describe()
    return tostring(self.name) .. ": " .. tostring(self.price)
end

function Product:shortName()
    return self.name
end


ClassFeaturesDemo = {}
ClassFeaturesDemo.__index = ClassFeaturesDemo

function ClassFeaturesDemo:new()
    local self = setmetatable({}, ClassFeaturesDemo)
    return self
end


function demoTypeWhen(x)
    return (function()
        if type(x) == "string" then
            return "string"
        elseif type(x) == "number" then
            return "number"
        elseif type(x) == "boolean" then
            return "boolean"
        else
            return "other"
        end
    end)()
end

function demoCounter()
    local c = Counter:new()
    c:increment()
    c:increment()
    c:increment()
    return c:readCount()
end

